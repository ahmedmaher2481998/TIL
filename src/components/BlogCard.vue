<script setup lang="ts">
import { TagChip } from '@/components'
import { type BlogType } from '@/types/index'
import { Card, CardDescription, CardFooter, CardHeader, CardTitle } from '@/components/ui/card'
import { formatDisplayDate, getInitials } from '@/utils'
import { Avatar, AvatarFallback, AvatarImage } from '@/components/ui/avatar'
import { CalendarTodayOutlined, ArrowUpwardOutlined } from '@vicons/material'
import { ref } from 'vue'
const { blog } = defineProps<{ blog: BlogType }>()
const ago = ref(formatDisplayDate(blog.createdat))
</script>

<template lang="html">
  <Card
    class="bg-green-100 text-white max-w-screen-lg relative overflow-hidden max-h-[500px] mx-2 md:mx-4"
  >
    <img class="bg-cover w-full rounded-lg" :src="blog.image" alt="{{ blog..title }}" />
    <div class="absolute bottom-0 glassy-feel w-full">
      <CardHeader>
        <!-- title with tldr  -->
        <div class="flex justify-between items-center w-full h-auto">
          <CardTitle class="font-semibold">{{ blog.title }}</CardTitle>
          <span
            class="h-10 text-sm w-10 rotate-45 hover:translate-x-1 hover:-translate-y-1 transition"
          >
            <ArrowUpwardOutlined class="font-thin" />
          </span>
        </div>
        <CardDescription class="text-slate-100 truncate">{{ blog.tldr }}</CardDescription>
      </CardHeader>

      <CardFooter class="flex items-center text-sm justify-between">
        <!-- author avatar with name  -->
        <div class="flex items-center justify-start">
          <div class="flex items-center justify-start gap-3">
            <Avatar>
              <AvatarImage :src="blog.user.avatar" :alt="blog.user.name" />
              <AvatarFallback>
                <!-- get initials  -->
                {{ getInitials(blog?.user.name ?? 'user name') }}
              </AvatarFallback>
            </Avatar>
            <span>{{ blog.user.name }}</span>
          </div>
          <div class="flex justify-center items-center gap-3 ms-4">
            <div class="h-10 p-2 border rounded-full w-10">
              <CalendarTodayOutlined />
            </div>
            <span> {{ ago }} </span>
          </div>
        </div>
        <!-- tags  -->
        <div class="flex gap-2">
          <TagChip
            v-for="tag in blog.tags"
            variant="glassy"
            :name="tag.title"
            :click="{
              href: '/tags/' + tag.id,
              linked: true
            }"
            :key="tag.id"
          />
        </div>
      </CardFooter>
    </div>
  </Card>
</template>

<style lang=""></style>
