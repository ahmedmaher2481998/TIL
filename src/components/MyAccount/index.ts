export { default as BlogsReadByUser } from './BlogsReadByUser.vue'
export { default as BlogsCreatedByUser } from './BlogsCreatedByUser.vue'
export { default as BlogCardSkeleton } from './BlogCard.Skeleton.vue'
export { default as UserProfile } from './UserProfile.vue'