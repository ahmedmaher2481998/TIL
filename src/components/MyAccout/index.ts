export { default as BlogsReadByUser } from './BlogsReadByUser.vue'
export { default as BlogsCreatedByUser } from './BlogsCreatedByUser.vue'
export { default as UserProfile } from './UserProfile.vue'