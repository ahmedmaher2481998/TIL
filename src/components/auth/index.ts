export { default as LoginPopUp } from './login/logInPopUp.vue'
export { default as LogoutPopUp } from './logout/logoutPopUp.vue'
export { default as RegisterPopUp } from './register/RegisterPopUp.vue'
export { default as UserMenu } from './userMenu/UserMenu.vue'
export { default as UserAvatarDisplay } from './userMenu/UserAvatarDisplay.vue'
export { default as LoginWithGoogleComponent } from './OAuth/logInWithGoogle.component.vue'
