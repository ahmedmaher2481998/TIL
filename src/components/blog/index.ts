export { default as BlogCard } from './BlogCard.vue'
export { default as TagChip } from './TagChip.vue'
export { default as FormInputField } from './createNewBlog/FormInputField.vue'
export { default as TagSelector } from './createNewBlog/TagSelector.vue'
export { default as CreateNewTagPopOver } from './createNewBlog/CreateNewTag.vue'
export { default as MarkDownContentInput } from './createNewBlog/MarkDownContentInput.vue'
export { default as TagChipComponent } from './DisplaySingleBlog/TagChip.component.vue'
