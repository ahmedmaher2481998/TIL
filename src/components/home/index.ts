export { default as FeaturedSectionComponent } from './FeaturedBlogSection/FeaturedSection.component.vue'
export { default as MainFeaturedBlogComponent } from './FeaturedBlogSection/MainFeaturedBlog.component.vue'
export { default as MainFeatureBlogSkeleton } from './FeaturedBlogSection/MainFeatureBlog.skeleton.vue'
export { default as SecondaryFeaturedBlogComponent } from './FeaturedBlogSection/SecondaryFeaturedBlog.component.vue'
export { default as SecondaryFeaturedBlogSkeleton } from './FeaturedBlogSection/SecondaryFeaturedBlog.skeleton.vue'
export { default as BlogsGrid } from './blogsGrid/BlogsGrid.vue'
export { default as BlogsGridSkeleton } from './blogsGrid/BlogsGrid.skeleton.vue'
