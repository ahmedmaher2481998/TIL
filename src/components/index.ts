// Lib components
export * from './ui/avatar'
export * from './ui/badge'
export { default as Button } from './ui/button/Button.vue'
export * from './ui/button'
export * from './ui/card'
export * from './ui/avatar'
export * from './ui/badge'
export * from './ui/dialog'
export * from './ui/input'
export * from './ui/label'
export * from './ui/command'
export * from './ui/form'
export * from './ui/popover'
export * from './ui/toast'
export * from './ui/skeleton'
export * from './ui/menubar'
export * from './ui/dropdown-menu'
// Main components [Shared ]
export { default as BlogCard } from './BlogCard.vue'
export { default as TagChip } from './TagChip.vue'
export { default as NavBar } from './NavBar.vue'
export { default as PopUp } from './PopUp.vue'
export { default as FormInputField } from './createNewBlog/FormInputField.vue'
export { default as ImageUpload } from './ImageUpload.vue'
// Create New Blog Components
export { default as TagSelector } from './createNewBlog/TagSelector.vue'
export { default as CreateNewTagPopOver } from './createNewBlog/CreateNewTag.vue'
export { default as MarkDownContentInput } from './createNewBlog/MarkDownContentInput.vue'

// auth components
export { default as LoginPopUp } from './auth/login/logInPopUp.vue'
export { default as LogoutPopUp } from './auth/logout/logoutPopUp.vue'
export { default as RegisterPopUp } from './auth/register/RegisterPopUp.vue'
export { default as UserMenu } from './auth/UserMenu.vue'
export { default as UserAvatarDisplay } from './auth/UserAvatarDisplay.vue'
