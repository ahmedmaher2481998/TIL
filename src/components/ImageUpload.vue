<script setup lang="ts">
import { Label, Input, FormItem } from '@/components'
import { UploadOutlined } from '@vicons/material'
import { ref, watch, toRefs, defineProps, withDefaults } from 'vue'
import { useField } from 'vee-validate'
import type { ClassValue } from 'clsx'
import clsx from 'clsx'

type propsType = {
  title: string
  name: string
  displayClasses?: ClassValue
}
const props = withDefaults(defineProps<propsType>(), {
  displayClasses: ''
})
const { title, name } = toRefs(props)

const { handleBlur, handleChange, errorMessage, value } = useField<File | null>(name, undefined, {
  initialValue: null
})
const blogCover = ref<HTMLImageElement | null>(null)
// display image when it's selected
watch(value, () => {
  if (value.value) {
    const reader = new FileReader()
    reader.onload = function (e: ProgressEvent<FileReader>) {
      blogCover!.value!.src = (e?.target?.result as string) ?? ''
    }
    reader.readAsDataURL(value.value)
  }
})
</script>
<template lang="html">
  <FormItem class="grid justify-start w-full items-center gap-1.5">
    <Label :for="name" class="mb-2 bg-" :class="{ 'text-red-600': Boolean(errorMessage) }">
      {{ title }}</Label
    >
    <label
      class="relative w-auto flex border p-2 rounded-lg cursor-pointer items-center"
      :for="name"
    >
      <span v-if="!value" class="flex items-center justify-center px-2">
        <UploadOutlined class="size-6 text-muted-foreground" />
      </span>
      <span class="flex flex-col">
        <img
          v-if="value"
          src=""
          ref="blogCover"
          class="shadow-lg rounded-lg object-cover"
          :class="displayClasses ? displayClasses : 'w-full h-auto'"
        />
        <span class="text-sm truncate text-gray-600 mt-4 text-center">
          {{ value ? value.name : 'Choose image' }}
        </span>
      </span>

      <Input
        accept="image/*"
        :id="name"
        type="file"
        class="pl-10 cursor-pointer hidden bg-green-600"
        @change="handleChange"
        @blur="handleBlur"
      />
    </label>
    <p v-if="Boolean(errorMessage)" class="text-sm font-medium text-destructive">
      {{ errorMessage }}
    </p>
  </FormItem>
</template>
