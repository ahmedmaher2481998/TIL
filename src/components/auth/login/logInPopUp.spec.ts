import { mount, VueWrapper } from '@vue/test-utils'
import { describe, afterEach, beforeEach, it, expect, vi } from 'vitest'
import { ref, Teleport } from 'vue'
import { LoginPopUp, Dialog } from '@/components'
import { useAuth } from '@/stores'
const logInMock = vi.fn().mockResolvedValueOnce(null) // Mock the login function
vi.mock('@/stores', () => ({
  useAuth: vi.fn(() => ({
    login: logInMock
  }))
}))

describe('Testing LoginPopUp Component ', () => {

  // let wrapper: VueWrapper
  // beforeEach(async () => {
  //   wrapper = await mount(LoginPopUp)
  // })
  // afterEach(async () => {
  //   // document.body.removeChild(teleportTarget);
  //   await wrapper?.unmount()
  // })
  console.log('start')
  const wrapper = mount(LoginPopUp)
  console.log('content', wrapper.html(), '--------------end here ----------')
  it('renders correctly', async () => {
    // expect(wrapper.findComponent(LoginPopUp).exists()).toBe(true)

    // expect(wrapper.find('input[name="email"]').exists()).toBe(true)

    // expect(wrapper.find('input[name="password"]').exists()).toBe(true)

    // expect(wrapper.find('button[type="submit"]').exists()).toBe(true)
  })

  // it('toggles password visibility', async () => {

  //   // Initially, the password field should be of type 'password'
  //   let passwordInput = wrapper.find('input[type="password"]')
  //   expect(passwordInput.exists()).toBe(true)

  //   // Click the toggle button to show password
  //   await wrapper.findComponent({ name: 'Eye' }).trigger('click')

  //   // Now, the password field should be of type 'text'
  //   passwordInput = wrapper.find('input[type="text"]')
  //   expect(passwordInput.exists()).toBe(true)
  // })

  // it('displays login error message on failed login', async () => {
  //   // const loginMock = useAuth().login
  //   // loginMock.mockRejectedValueOnce(new Error('Invalid credentials'))


  //   // Fill in the form
  //   await wrapper.find('input[type="email"]').setValue('wrong-email@example.com')
  //   await wrapper.find('input[type="password"]').setValue('wrongpassword')

  //   // Submit the form
  //   await wrapper.find('form').trigger('submit.prevent')

  //   // Wait for the login process to complete
  //   await wrapper.vm.$nextTick()

  //   // Expect the login error message to appear
  //   expect(wrapper.find('.text-destructive').text()).toContain('Invalid credentials')
  // })

  // it('submits form successfully when correct credentials are provided', async () => {
  //   const loginMock = useAuth().login


  //   // Fill in the form with valid credentials
  //   await wrapper.find('input[type="email"]').setValue('user@gmail.com')
  //   await wrapper.find('input[type="password"]').setValue('123123123')

  //   // Submit the form
  //   await wrapper.find('form').trigger('submit.prevent')

  //   // Ensure login function is called with correct credentials
  //   expect(loginMock).toHaveBeenCalledWith({
  //     email: 'user@gmail.com',
  //     password: '123123123'
  //   })
  // })
})
