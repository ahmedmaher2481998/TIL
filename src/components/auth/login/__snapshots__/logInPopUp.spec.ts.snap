// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`testing login pop up component > matches the snapshot 1`] = `
"<form keep-values="" class="w-full">
  <div class="w-full space-y-3 mb-4">
    <!--v-if-->
    <div>
      <div data-v-87b53926="" class="space-y-2"><label data-v-87b53926="" for="radix-v-0-form-item" class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-secondary-foreground">Email *</label>
        <div data-v-87b53926="" class="relative w-full items-center"><input data-v-87b53926="" class="flex text-foreground bg-input h-10 w-full rounded-md border border-input px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" autocomplete="off" type="email" required="" placeholder="user@email.com" name="email" id="radix-v-0-form-item" aria-describedby="radix-v-0-form-item-description" aria-invalid="false"></div>
        <!--v-if-->
        <!---->
      </div>
    </div>
    <div class="">
      <div data-v-87b53926="" class="space-y-2"><label data-v-87b53926="" for="radix-v-1-form-item" class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-secondary-foreground">password *</label>
        <div data-v-87b53926="" class="relative w-full items-center"><input data-v-87b53926="" class="flex text-foreground bg-input h-10 w-full rounded-md border border-input px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 pr-10" autocomplete="off" type="password" required="" placeholder="*****" name="password" id="radix-v-1-form-item" aria-describedby="radix-v-1-form-item-description" aria-invalid="false"><span class="absolute cursor-pointer end-0 inset-y-0 flex items-center justify-center px-2"><transition-stub name="fade" mode="out-in" appear="false" persisted="false" css="true"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-eye-icon size-6 text-foreground"><path d="M2 12s3-7 10-7 10 7 10 7-3 7-10 7-10-7-10-7Z"></path><circle cx="12" cy="12" r="3"></circle></svg></transition-stub></span></div>
        <!--v-if-->
        <!---->
      </div>
    </div>
  </div>
  <div class="flex items-center justify-center gap-3 flex-col"><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full" type="submit"> login </button>
    <div data-orientation="horizontal" role="separator" label="Or" class="shrink-0 bg-border relative h-px w-full"><span class="text-xs text-secondary absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex justify-center items-center h-[1px] py-1 px-2">Or</span></div>
    <!---->
  </div>
</form>"
`;
